### ISSUE TITLE
IR - CPointer

### ISSUE THREAD

**Issue by CalebLItalien** [UnionCompilerDesign/compiler_core#69]
### Details:
Introduce an `IRPointer` type within the `IRGenerator` module to encapsulate and manage raw `LLVMValueRef` pointers safely. 

### Issue Type:
- [x] Feature
- [ ] Bugfix
- [ ] Documentation
- [ ] Refactoring
- [ ] Other

### Requirements:
- Wrap `LLVMValueRef` to hide direct access to raw pointers
     - Common operations for access, setting, and dropping
     - Null checks
     - Block double frees
     - No use after free
- `Debug`, `Clone`, `Copy`, `Hash`

### Notes:

**Comment by CalebLItalien** [comment]
Need to be covered
- LLVMValueRef
- LLVMBasicBlockRef
- LLVMContextRef
- LLVMModuleRef
- LLVMBuilderRef
- LLVMTypeRef

**Comment by CalebLItalien** [comment]
- Represent LLVM objects with unique ids. Use these identifiers to interact with the objects outside of llvm_lib. 

```rust
/// --- Handle Types --- ///
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub struct ValueHandle(usize);

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub struct BasicBlockHandle(usize);
...

/// --- Registry --- ///
struct LLVMRegistry {
    values: HashMap<ValueHandle, LLVMValueRef>,
    basic_blocks: HashMap<BasicBlockHandle, LLVMBasicBlockRef>,
    ...
    next_handle: usize, // Generates unique IDs
}

impl LLVMRegistry {
    fn new() -> Self {
        Self {
            values: HashMap::new(),
            basic_blocks: HashMap::new(),
            next_handle: 0,
        }
    }

    fn get_value(&self, handle: ValueHandle) -> Option<&LLVMValueRef> {
        self.values.get(&handle)
    }

    fn create_value(&mut self, value: LLVMValueRef) -> ValueHandle {
        let handle = ValueHandle(self.next_handle);
        self.next_handle += 1;
        self.values.insert(handle, value);
        handle
    }
    ...
}
```

**Comment by CalebLItalien** [comment]
```rust
#[derive(Debug, Clone, Copy)]
struct IRPointer {
    ptr: *mut llvm_sys::prelude::LLVMValueRef, // Use the appropriate LLVM type
}

impl IRPointer {
    fn new(ptr: *mut llvm_sys::prelude::LLVMValueRef) -> Self {
        assert!(!ptr.is_null(), "Attempting to create IRPointer with null pointer");
        IRPointer { ptr }
    }

    /// Provides safe access to the underlying LLVM pointer.
    fn as_ref(&self) -> &llvm_sys::prelude::LLVMValueRef {
        unsafe { &*self.ptr }
    }
}

impl Deref for IRPointer {
    type Target = llvm_sys::prelude::LLVMValueRef;

    fn deref(&self) -> &Self::Target {
        self.as_ref()
    }
}
```

**Comment by CalebLItalien** [comment]
This project has been moved to `safe_llvm`.

**Comment by CalebLItalien** [comment]
This has been updated to `CPointer`. It is integrated but remains in testing until a revised testing approach is taken to ensure functionality of `safe_llvm`.

**Comment by CalebLItalien** [comment]
This issue will be closed with #123.

**Comment by CalebLItalien** [comment]
Closing this issue, since it was merged into `backend` without review. The `backend` remains unstable and will need to be revisited in #123.
